contracts/exchange-wrappers/ZeroExV2MultiOrderExchangeWrapper.sol

```diff
@@ -0,0 +1,399 @@
+import { MathHelpers } from "../lib/MathHelpers.sol";
+import { TokenInteract } from "../lib/TokenInteract.sol";
+
+
+/**
+ * @title ZeroExV2MultiOrderExchangeWrapper
+ * @author dYdX
+ *
+ * dYdX ExchangeWrapper to interface with 0x Version 2. Sends multiple orders at once. Assumes no
+ * fees.
+ */
+contract ZeroExV2MultiOrderExchangeWrapper is
+    LibFillResults,
+    LibOrder,
+    ExchangeWrapper
+{
+    using SafeMath for uint256;
+    using TokenInteract for address;
+
+    // ============ Constants ============
+
+    // number of bytes in the maxPrice data
+    uint256 constant PRICE_DATA_LENGTH = 64;
+
+    // number of bytes per (order+signature)
+    uint256 constant ORDER_DATA_LENGTH = 322;
+
+    // ============ Structs ============
+
+    struct TokenAmounts {
+        uint256 takerAmount;
+        uint256 makerAmount;
+    }
+
+    // ============ State Variables ============
+
+    // address of the ZeroEx V2 Exchange
+    address public ZERO_EX_EXCHANGE;
+
+    // address of the ZeroEx V2 ERC20Proxy
+    address public ZERO_EX_TOKEN_PROXY;
+
+    // ============ Constructor ============
+
+    constructor(
+        address zeroExExchange,
+        address zeroExProxy
+    )
+        public
+    {
+        ZERO_EX_EXCHANGE = zeroExExchange;
+        ZERO_EX_TOKEN_PROXY = zeroExProxy;
+    }
+
+    // ============ Public Functions ============
+
+    function exchange(
+        address /* tradeOriginator */,
+        address receiver,
+        address makerToken,
+        address takerToken,
+        uint256 requestedFillAmount,
+        bytes calldata orderData
+    )
+        external
+        returns (uint256)
+    {
+        // parse all order data
+        validateOrderData(orderData);
+        TokenAmounts memory priceRatio = parseMaxPriceRatio(orderData);
+        Order[] memory orders = parseOrders(orderData, makerToken, takerToken);
+        bytes[] memory signatures = parseSignatures(orderData);
+
+        // ensure that the exchange can take the tokens from this contract
+        ensureAllowance(
+            takerToken,
+            ZERO_EX_TOKEN_PROXY,
+            requestedFillAmount
+        );
+
+        // do the exchange
+        FillResults memory totalFillResults = IExchange(ZERO_EX_EXCHANGE).marketSellOrdersNoThrow(
+            orders,
+            requestedFillAmount,
+            signatures
+        );
+
+        // validate that all taker tokens were sold
+        require(
+            totalFillResults.takerAssetFilledAmount == requestedFillAmount,
+            "ZeroExV2MultiOrderExchangeWrapper#exchange: Cannot sell enough taker token"
+        );
+
+        // validate that max price is not violated
+        validateTradePrice(
+            priceRatio,
+            totalFillResults.takerAssetFilledAmount,
+            totalFillResults.makerAssetFilledAmount
+        );
+
+        // set allowance
+        ensureAllowance(makerToken, receiver, totalFillResults.makerAssetFilledAmount);
```

Comment:
So this means we could have allowance set on arbitrary addresses. This is ok because this contract only transiently holds funds.  I’m trying to think of an attack where an attacker could steal funds that are here transiently. For solo, it’d be if an attacker contract got control of the runtime anytime between when solo transfers the EW the funds and this function returns.

I don’t think this is possible because it’s like:

- solo transfers funds
- solo calls EW
- EW calls to set allowance on 0x (potentially)
- EW calls 0x
- 0x calls token contracts to transfer
- EW sets allowance on solo (maybe)
- return

So as long as we trust solo, token contracts, 0x should be fine. Seem right?

Reply:
Yep, same as all other ExchangeWrappers, use bad tokens (or bad exchanges), win bad prizes. Also that we explicitly ensure token balances when using `transferFrom` afterwards from Solo or whatever other contract is using this ExchangeWrapper